File: dart/lib/src/models/asset.dart
import 'amount.dart';

/// An asset (native currency or token) with balance information.
class Asset {
  /// Server-assigned identifier.
  final String id;

  /// Canonical asset identifier in `"<type>.<chainId>.<address>"`
  /// form, e.g. `"evm.1.NATIVE"`, `"solana.mainnet.EPjFW…"`,
  /// `"evm.1.0xA0b86991…"`. Native assets end with `.NATIVE`; tokens
  /// end with the on-chain contract / mint address. Use [isNative]
  /// to branch.
  final String key;

  /// Human-readable asset name (e.g. `"Ethereum"`).
  final String name;

  /// Ticker symbol (e.g. `ETH`, `BTC`).
  final String symbol;

  /// Current balance of this asset.
  final Amount amount;

  /// Coarse-grained asset family — currently `"fungible"` for every
  /// runtime-emitted asset (native currencies + ERC-20 / SPL-token
  /// balances both). Reserved for a future `"nft"` distinction.
  ///
  /// **Don't branch native-vs-token on this field** — use
  /// [isNative] instead. Native is signalled by the `.NATIVE`
  /// suffix on [key], not by [type].
  final String type;

  /// Network ID this asset belongs to.
  final String network;

  /// Whether this asset is on a test network.
  final bool testNet;

  /// Balance converted to fiat currency.
  final Amount? fiatAmount;

  /// ISO 4217 fiat currency code for [fiatAmount].
  final String? fiatCurrency;

  /// Market price quote data for this asset.
  final FiatQuote? fiatQuote;

  /// Additional metadata about the asset.
  final Map<String, dynamic>? info;

  /// Timestamp when the asset was first seen.
  final DateTime created;

  /// Timestamp when the asset was last updated.
  final DateTime updated;

  const Asset({
    required this.id,
    required this.key,
    required this.name,
    required this.symbol,
    required this.amount,
    required this.type,
    required this.network,
    required this.testNet,
    this.fiatAmount,
    this.fiatCurrency,
    this.fiatQuote,
    this.info,
    required this.created,
    required this.updated,
  });

  /// True when this Asset is the chain's native currency (ETH / SOL /
  /// BTC / etc.) — i.e. when [key] ends with `.NATIVE` (libwallet's
  /// canonical native sentinel; see `wltnet/network.go`).
  ///
  /// Use this rather than matching on [type] / [symbol] / [name]:
  /// libwallet emits `Type: "fungible"` for native AND tokens, the
  /// symbol can collide (a token can ship as "ETH" too), and the
  /// name is localised in some hosts. The `.NATIVE` suffix on the
  /// key is the only invariant.
  bool get isNative => key.endsWith('.NATIVE');

  /// The on-chain mint / contract part of [key], or `null` for a
  /// native asset. For `"evm.1.0xA0b86991…"` returns
  /// `"0xA0b86991…"`; for `"solana.mainnet.EPjFW…"` returns
  /// `"EPjFW…"`; for `"evm.1.NATIVE"` returns `null`.
  String? get tokenAddress {
    if (isNative) return null;
    final idx = key.lastIndexOf('.');
    if (idx < 0 || idx == key.length - 1) return null;
    return key.substring(idx + 1);
  }

  factory Asset.fromJson(Map<String, dynamic> json) {
    return Asset(
      id: json['Id'] as String? ?? json['key'] as String? ?? '',
      key: json['key'] as String? ?? '',
      name: json['name'] as String? ?? json['Name'] as String? ?? '',
      symbol: json['symbol'] as String? ?? json['Symbol'] as String? ?? '',
      amount: Amount.fromJson(json['amount'] ?? json['Amount']),
      type: json['type'] as String? ?? json['Type'] as String? ?? '',
      network: json['network'] as String? ?? json['Network'] as String? ?? '',
      testNet: json['testnet'] as bool? ?? json['TestNet'] as bool? ?? false,
      fiatAmount: json['fiat_amount'] != null
          ? Amount.fromJson(json['fiat_amount'])
          : json['FiatAmount'] != null
              ? Amount.fromJson(json['FiatAmount'])
              : null,
      fiatCurrency: json['fiat_currency'] as String? ??
          json['FiatCurrency'] as String?,
      fiatQuote: json['fiat_quote'] != null
          ? FiatQuote.fromJson(json['fiat_quote'])
          : json['FiatQuote'] != null
              ? FiatQuote.fromJson(json['FiatQuote'])
              : null,
      info: json['info'] as Map<String, dynamic>?,
      created: DateTime.parse(
          json['Created'] as String? ?? DateTime.now().toIso8601String()),
      updated: DateTime.parse(
          json['Updated'] as String? ?? DateTime.now().toIso8601String()),
    );
  }
}

/// Price quote data from CoinMarketCap or similar.
class FiatQuote {
  /// Current price per unit in the quote currency.
  final double price;

  /// 24-hour trading volume.
  final double volume24h;

  /// 24-hour change in trading volume (percentage).
  final double volumeChange24h;

  /// 24-hour price change (percentage).
  final double percentChange24h;

  /// 1-hour price change (percentage).
  final double percentChange1h;

  /// Total market capitalization.
  final double marketCap;

  /// Timestamp of the last price update.
  final DateTime lastUpdate;

  const FiatQuote({
    required this.price,
    required this.volume24h,
    required this.volumeChange24h,
    required this.percentChange24h,
    required this.percentChange1h,
    required this.marketCap,
    required this.lastUpdate,
  });

  factory FiatQuote.fromJson(dynamic json) {
    return FiatQuote(
      price: (json['price'] as num).toDouble(),
      volume24h: (json['volume_24h'] as num).toDouble(),
      volumeChange24h: (json['volume_change_24h'] as num).toDouble(),
      percentChange24h: (json['percent_change_24h'] as num).toDouble(),
      percentChange1h: (json['percent_change_1h'] as num).toDouble(),
      marketCap: (json['market_cap'] as num).toDouble(),
      lastUpdate:
          DateTime.tryParse(json['last_updated'] ?? '') ?? DateTime.now(),
    );
  }
}

File: rust/src/models/asset.rs
//! wltasset — asset/balance rows. Port of the Go `wltasset` package.
//!
//! Read surface only: assets are populated by balance discovery (RPC), which
//! lands with wltnet; there is no create endpoint. Unlike the other models,
//! Asset uses lowercase JSON keys (explicit Go `json:` tags) — except Created/
//! Updated, which have none and stay PascalCase. Amount is stored as its
//! `{v,e,f}` JSON and round-trips through [`crate::Amount`].

use serde::{Deserialize, Serialize};
use crate::{Amount, Env, Result, SqlValue};

const TABLE_DDL: &str = r#"CREATE TABLE IF NOT EXISTS "Asset" ("Id" text, "Key" text, "Name" text, "Symbol" text, "Amount" text, "Type" text, "Network" text, "Created" text, "Updated" text, PRIMARY KEY ("Id"));
CREATE UNIQUE INDEX IF NOT EXISTS "Asset_Key" ON "Asset" ("Key");"#;
const COLS: &str = r#""Id", "Key", "Name", "Symbol", "Amount", "Type", "Network", "Created", "Updated""#;

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct Asset {
    #[serde(rename = "id", default, skip_serializing_if = "String::is_empty")]
    pub id: String,
    #[serde(rename = "key", default)]
    pub key: String,
    #[serde(rename = "name", default)]
    pub name: String,
    #[serde(rename = "symbol", default)]
    pub symbol: String,
    #[serde(rename = "amount")]
    pub amount: Amount,
    #[serde(rename = "type", default)]
    pub kind: String,
    #[serde(rename = "network", default, skip_serializing_if = "String::is_empty")]
    pub network: String,
    #[serde(rename = "Created", default)]
    pub created: String,
    #[serde(rename = "Updated", default)]
    pub updated: String,

    // Computed (never persisted) fiat conversion, matching Go's sql:"-" fields.
    #[serde(rename = "fiat_amount", default, skip_serializing_if = "Option::is_none")]
    pub fiat_amount: Option<Amount>,
    #[serde(rename = "fiat_currency", default, skip_serializing_if = "String::is_empty")]
    pub fiat_currency: String,
    #[serde(rename = "fiat_quote", default, skip_serializing_if = "Option::is_none")]
    pub fiat_quote: Option<crate::quote::CmcQuoteEntry>,
    #[serde(rename = "testnet", default, skip_serializing_if = "std::ops::Not::not")]
    pub testnet: bool,
}

impl Asset {
    /// Native currency? (`.NATIVE` suffix on Key — Go `Asset.IsNative`).
    pub fn is_native(&self) -> bool {
        self.key.ends_with(".NATIVE")
    }

    /// Port of Go `Asset.ConvertTo`: multiply the balance by the token's fiat
    /// price (from the quote table) and populate the fiat_* fields. No-op for
    /// testnet assets. `Ok(false)` when no quote/currency entry exists.
    pub fn convert_to(&mut self, env: &Env, currency: &str) -> Result<bool> {
        if self.testnet {
            return Ok(false);
        }
        let quote = match crate::quote::get_quotes_for_token(env, &self.symbol)? {
            Some(q) => q,
            None => return Ok(false),
        };
        let entry = match quote.quote.get(currency) {
            Some(e) => e.clone(),
            None => return Ok(false),
        };
        // price as an Amount (8 decimals — "more decimals always good"), then
        // fiat = balance * price.
        let price = Amount::from_float64(entry.price, 8);
        let mut fiat = Amount::new(0, 8);
        fiat.mul(&self.amount, &price);
        self.fiat_amount = Some(fiat);
        self.fiat_currency = currency.to_owned();
        self.fiat_quote = Some(entry);
        Ok(true)
    }
}

pub fn init(env: &Env) -> Result<()> {
    env.ensure_table(TABLE_DDL)
}

pub fn fetch(env: &Env, id: &str) -> Result<Option<Asset>> {
    let sql = format!(r#"SELECT {COLS} FROM "Asset" WHERE "Id" = ?1"#);
    let rows = env.query(&sql, vec![SqlValue::Text(id.to_owned())])?;
    Ok(rows.first().map(|r| row_to_asset(r)))
}

pub fn list(env: &Env) -> Result<Vec<Asset>> {
    let sql = format!(r#"SELECT {COLS} FROM "Asset" ORDER BY "Key" ASC"#);
    let rows = env.query(&sql, Vec::new())?;
    let mut assets: Vec<Asset> = rows.iter().map(|r| row_to_asset(r)).collect();
    // Registered ERC-20 tokens on the current EVM network, with live balances
    // (port of the EVM leg Go `computeAssets` gained in wltbase/asset.go).
    // Appended after the persisted rows; best-effort, so a missing current
    // network/account or an unresolvable RPC simply contributes nothing.
    assets.extend(registered_erc20_assets(env));
    Ok(assets)
}

/// The current EVM network's registered ERC-20 tokens as live-balance assets
/// (port of the EVM leg of Go `computeAssets`). Unlike Solana — where token
/// accounts are enumerable on-chain — EVM has no cheap owner→tokens query, so
/// the user's Token registry (Token:create / Token:discoverToken / swap
/// EnsureToken) is the source of truth. Zero balances are included on purpose:
/// a token the user explicitly registered shows as "0" rather than vanishing.
/// A per-token RPC failure skips that token instead of failing the whole list.
/// Best-effort throughout — any setup failure yields no rows.
fn registered_erc20_assets(env: &Env) -> Vec<Asset> {
    let net = match crate::models::network::fetch(env, "@") {
        Ok(Some(n)) if n.kind == "evm" => n,
        _ => return Vec::new(),
    };
    let account = match crate::models::account::current(env) {
        Ok(Some(a)) if a.kind == "ethereum" && !a.address.is_empty() && a.address != "N/A" => a,
        _ => return Vec::new(),
    };
    let rpc = match net.resolved_rpc() {
        Ok(r) => r,
        Err(_) => return Vec::new(),
    };
    let tokens = match crate::models::token::tokens_by_network(env, &net.id) {
        Ok(t) => t,
        Err(_) => return Vec::new(),
    };
    let mut out = Vec::new();
    for token in tokens {
        if token.kind != "erc20" {
            continue;
        }
        let bal = match crate::erc20::balance_of(&rpc, &token.address, &account.address) {
            Ok(b) => b,
            Err(_) => continue, // per-token RPC failure: skip (Go logs + continues)
        };
        // Fall back to a truncated contract address when the row carries no
        // display metadata (matches Go's name/symbol fallbacks).
        let name = if token.name.is_empty() {
            format!("{}...", &token.address[..token.address.len().min(10)])
        } else {
            token.name.clone()
        };
        let symbol = if token.symbol.is_empty() {
            token.address[..token.address.len().min(8)].to_owned()
        } else {
            token.symbol.clone()
        };
        out.push(Asset {
            id: String::new(),
            key: format!("{}.{}", net.key_prefix(), token.address),
            name,
            symbol,
            amount: Amount::new_raw(bal, token.decimals),
            kind: "fungible".to_owned(),
            network: net.id.clone(),
            created: String::new(),
            updated: String::new(),
            fiat_amount: None,
            fiat_currency: String::new(),
            fiat_quote: None,
            testnet: net.testnet,
        });
    }
    out
}

fn row_to_asset(row: &[SqlValue]) -> Asset {
    let text = |i: usize| row.get(i).and_then(|v| v.as_text()).unwrap_or("").to_owned();
    let amount = row
        .get(4)
        .and_then(|v| v.as_text())
        .and_then(|s| serde_json::from_str::<Amount>(s).ok())
        .unwrap_or_else(|| Amount::new(0, 0));
    Asset {
        id: text(0),
        key: text(1),
        name: text(2),
        symbol: text(3),
        amount,
        kind: text(5),
        network: text(6),
        created: text(7),
        updated: text(8),
        fiat_amount: None,
        fiat_currency: String::new(),
        fiat_quote: None,
        testnet: false,
    }
}
